presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -438,13 +439,40 @@ protected TupleDescriptor visitInsert(Insert insert, AnalysisContext context)
         accessControl.checkCanInsertIntoTable(session.getIdentity(), targetTable);
         analysis.setInsertTarget(targetTableHandle.get());
 
-        List<ColumnMetadata> columns = metadata.getTableMetadata(session, targetTableHandle.get()).getColumns();
-        Iterable<Type> tableTypes = columns.stream()
-                .filter(column -> !column.isHidden())
-                .map(ColumnMetadata::getType)
-                .collect(toImmutableList());
+        List<ColumnMetadata> allTableColumns = metadata.getTableMetadata(session, targetTableHandle.get()).getColumns();
+        final Iterable<Type> tableTypes;
+        if (insert.getColumns().isPresent()) {
+            List<String> insertColumns = insert.getColumns().get();
+            analysis.setInsertColumns(insert.getColumns().get());
 
-        Iterable<Type> queryTypes = transform(descriptor.getVisibleFields(), Field::getType);
+            Set<String> columnNames = new HashSet<>();
+            for (String name : insertColumns) {
+                if (!columnNames.add(name)) {
+                    throw new SemanticException(DUPLICATE_COLUMN_NAME, insert, "Insert column name is specified more than once: %s", name);
```

Comment:
I think it is better to put the user string on the right side of the error when possible. This avoid issues with quoting, spaces, etc. that can appear in identifiers (which we don't handle at all now in error messages).